### DISCUSSION TITLE
Multiple file download not working after cefsharp upgrade

### DISCUSSION THREAD

**Discussion by aryapram** [cefsharp/CefSharp#5018] [General]
### Is there an existing issue for this?

- [X] I have searched both open/closed issues, no issue already exists.

### CefSharp Version

127.3.50

### Operating System

Windows 10

### Architecture

x64

### .Net Version

4.6.2

### Implementation

WinForms

### Reproduction Steps

After upgrading CefSharp from 114.2.120 to 127.3.50 the multiple file download is not working 

I am using the below angular code  to download the xml file. Need to download 3 files.
this.serviceObj.saveTextAsFile(report1, report1);
this.serviceObj.saveTextAsFile(report2, report2);
this.serviceObj.saveTextAsFile(report3, report13);
saveTextAsFile(textToWrite: string, fileNameToSaveAs: string) {

    var textFileAsBlob = new Blob([textToWrite], { type: 'text/xml' });
    var downloadLink = document.createElement("a");

    downloadLink.download = fileNameToSaveAs;

    if ((window as any).webkitURL != null) {

      // Chrome allows the link to be clicked
      // without actually adding it to the DOM.
      var textFile = (window as any).webkitURL.createObjectURL(textFileAsBlob);
      downloadLink.href = textFile;
      downloadLink.click();

      // manually revoke the object URL to avoid memory leaks.
      (window as any).webkitURL.revokeObjectURL(textFile);
      downloadLink.remove();
    }
 }

Below is my custom Download handler code 
    public bool OnBeforeDownload(IWebBrowser chromiumWebBrowser, IBrowser browser, DownloadItem downloadItem, IBeforeDownloadCallback callback)
    {
        OnBeforeDownloadFired?.Invoke(this, downloadItem);

        if (!callback.IsDisposed)
        {
            using (callback)
            {
                callback.Continue(downloadItem.SuggestedFileName, showDialog: true);
            }
        }
        return true;
    }
   
    public void OnDownloadUpdated(IWebBrowser chromiumWebBrowser, IBrowser browser, DownloadItem downloadItem, IDownloadItemCallback callback)
    {
        OnDownloadUpdatedFired?.Invoke(this, downloadItem);

        if (downloadItem.IsComplete || downloadItem.IsCancelled)
        {                

            if (isFileToBeOpenedByDefaultApplication(downloadItem.MimeType, browser))
            {
                if (browser.IsPopup)
                {
                    browser.CloseBrowser(false);
                }
                if (downloadItem.IsComplete)
                {
                    try
                    {
                        var dg = new Action(() =>
                        {
                            MessageBox.Show(string.Format("{0} has been downloaded successfully to {1}", Path.GetFileName(downloadItem.FullPath),
                                        Path.GetDirectoryName(downloadItem.FullPath)), "File Download");
                        });
                        System.Windows.Threading.Dispatcher.CurrentDispatcher.BeginInvoke(dg);

                        Process.Start(downloadItem.FullPath);  
                    }
                    catch (Exception e)
                    {
                        Logger.LogException("Process.Start Exception", e);
                    }                        
                }

            }
            else
            {
                if (downloadItem.IsComplete)
                {
                    var dg = new Action(() =>
                    {
                        MessageBox.Show(string.Format("{0} has been downloaded successfully to {1}", Path.GetFileName(downloadItem.FullPath),
                                    Path.GetDirectoryName(downloadItem.FullPath)), "File Download");
                    });
                    System.Windows.Threading.Dispatcher.CurrentDispatcher.BeginInvoke(dg);
                }
            }

        }

    }
      

But only one file is getting downloaded. after that the download itself is not working. Please provide a solution 

### Expected behavior

3 xml files need to be downloaded.

### Actual behavior

Only one file downloaded

### Regression?

worked on CefSharp 114.2.120 

### Known Workarounds

no work arounds found

### Does this problem also occur in the CEF Sample Application

Yes using WinForms command line args

### Other information

no log infomation

**Comment by amaitland** [comment]
Converting to discussion as this is unlikely a bug in `CefSharp`. 

There were major upstream changes recently. You can read about them in more detail at https://groups.google.com/g/cef-announce/c/s1WaovAopFo/m/LV5eiNX1BgAJStarting

Many behaviors have changed.

> What version are you using? Please only open an issue if you can reproduce the problem with version 130.1.90 or later.

As per https://github.com/cefsharp/CefSharp/blob/master/.github/ISSUE_TEMPLATE/bug_report.yml

Have you tested the current supported version? At time of writing `M131`?


If it doesn't work in `M131` then you can try switching to a `Chrome Styled` window. See https://github.com/cefsharp/CefSharp/issues/4835#issuecomment-2182373137

If that doesn't work then the best place to ask is on https://www.magpcss.org/ceforum/viewforum.php?f=6&sid=1eb1597a73e5c70e6211f420b9a10a60

**Reply by amaitland** [reply]
https://bitbucket.org/chromiumembedded/cef/commits/2ea7459a89fb maybe relevant

Did you check the log file for errors?

**Comment by max3163** [comment]
I've the same issue, since I use the Runtime Chrome with ALLOW style, i can't download multiple files. 
I switch for a test with chrome style, it's this popup  which block :

![image](https://github.com/user-attachments/assets/484a3801-0db1-4575-af31-e7f27efae888)

I asked on magpcss forum too for help to find a parameter or settings to disable this behaviour

**Reply by amaitland** [reply]
Suggested solution is https://magpcss.org/ceforum/viewtopic.php?f=6&t=19835#p55603

https://github.com/cefsharp/CefSharp/blob/f684da355f9b4dcf7f93c1be4d3d0fc92e98d2f5/CefSharp/IRequestContext.cs#L260

Example of `SetPreference`, which can be adapted to call `SetContentSetting` instead.

```
//To execute on the CEF UI Thread you can use 
Cef.UIThreadTaskFactory.StartNew(delegate
{
    string errorMessage;
    //Use this to check that settings preferences are working in your code

    //the browser variable is an instance of ChromiumWebBrowser
    var success = browser.RequestContext.SetPreference("webkit.webprefs.minimum_font_size", 24, out errorMessage);
}); 
```

**Reply by YyzzzM** [reply]
Have you solved this problem? Im having the same issue, and setting 'AutomaticDownloads' to 'Allow' doesnt seem to help either.

**Reply by campersau** [reply]
I haven't tried it but you can allow the permission in the permission handler: https://github.com/campersau/CefSharp/blob/updateerrorcodepermissionrequesttype/CefSharp%2FEnums%2FPermissionRequestType.cs#L32

**Comment by m-aljammali-stramit** [comment]
I am facing the same issue.


I am using Alloy Style. ( I think just started with .NET development less than a month ago so I am not sure about anything)
The CefSharp Chromium-based browser component (WinForms control). 137.0.100
and on Windows 10

My current work around: I read and edit the embedded browser preferences file and add URL to automatic_downloads

```
        Try
            Dim jsonData As String = File.ReadAllText(cachePath + "\Default\Preferences")
            Dim prefernceList = JsonConvert.DeserializeObject(jsonData)

            Dim nestedItem As JObject = New JObject()
            nestedItem("last_modified") = "13395737733446032"
            nestedItem("setting") = 1

            Dim parentObject As JObject =
            DirectCast(DirectCast(DirectCast(DirectCast(prefernceList("profile"), JObject)("content_settings"), JObject)("exceptions"), JObject)("automatic_downloads"), JObject)
            'parentObject("http://localhost:8081,*") = nestedItem
            parentObject(geturl() + ",*") = nestedItem
            ' Update json data string
            jsonData = JsonConvert.SerializeObject(prefernceList)
            System.IO.File.WriteAllText(cachePath + "\Default\Preferences", jsonData)
        Catch ex As Exception
        End Try

        settings.CachePath = cachePath
        CefSharp.Cef.Initialize(settings)
```

I noticed the suggested solution requires editing CefSharp package. not sure which work around is better. an advice will be appreciated.

**Reply by amaitland** [reply]
> I noticed the suggested solution requires editing CefSharp package. not sure which work around is better. an advice will be appreciated.

The suggested solution is to call `SetContentSetting`, no changes to the package are required for this.

**Reply by blyantspisser** [reply]
I thought the suggested solution looked complicated. Could you explain it a little bit more closely, maybe with an example? I am trying currently to roll it back to 114.2.120 which is the version that the OP says is still working as before, but I am having some trouble with that as well. It is starting to look like maintaining a proprietary browser for us is no longer in the cards.

I found 124.3.80 to be the latest version  I can use.

**Reply by blyantspisser** [reply]
What is the meaning of requestUrl and topLevelUrl? What is the best thing to set these addresses to in order to always be able to download multiple files? As long as all my pages are subpages within the same domain or "base address" I think I can set both strings to this base address and I'll be good. It looks like it is working. But what if I navigate to a different domain? I have to do another SetContentSetting? Does SetContentSetting add urls to a list? Does a setting exist  that means "always allow all sites"?

**Reply by amaitland** [reply]
https://cef-builds.spotifycdn.com/docs/142.0/classCefRequestContext.html#addd486f82451cd1cf050df0a107ecdcc
https://github.com/cefsharp/CefSharp/blob/cefsharp/141/CefSharp/IRequestContext.cs#L260

CEF documentation is fairly generic.

> Does a setting exist that means "always allow all sites"?

Did you try setting both urls as empty strings? As per the doc?

**Reply by blyantspisser** [reply]
The docs say that you set "default" by using empty strings. It was unclear to me what "default" means, but when I test it, it gives the same result as if I don't use a request handler at all, so I am guessing that default is the same thing  that happens if I don't call SetContentSetting at all.
Since the web page I am serving at the moment is a single contained page without any links that take you out of the domain, I can let the user configure the home address and I use that in both address fields in the call to SetContentSetting and this seems to work. I don't really know how SetContentSetting works, is it a whilelist of urls that can multi download, or is it one that is replaced if I set a new one? If I do have a site with external links, should I add any new urls that the user navigates to? I find it unclear. Not so simple anymore to make my own web browser.

**Comment by blyantspisser** [comment]
I made it work for my case where I have a fixed home url and I can allow multi file DL for everything under by using the RequestContext:

_browser = new ChromiumWebBrowser(HomeAddress);
var requestContextSettings = new RequestContextSettings();
var requestContext = new RequestContext(requestContextSettings);
_browser.RequestContext = requestContext;
Cef.UIThreadTaskFactory.StartNew(delegate
{
_browser.RequestContext.SetContentSetting(_homeAddress, _homeAddress, CefSharp.Enums.ContentSettingTypes.AutomaticDownloads, CefSharp.Enums.ContentSettingValues.Allow);
});


Now, if you have an application where you can navigate to some other domain, I expect you have to call SetContentSetting again to allow that other domain specifically. I am guessing that this creates a "whitelist" of web sites where multi file downloads are allowed, but I have only tested it for my own case where I have a fixed home address.
